controllers/model/model_controller.go

```diff
@@ -0,0 +1,192 @@
+/*
+Copyright 2020 The Alibaba Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/alibaba/kubedl/controllers/model/storage"
+	"github.com/go-logr/logr"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
+
+	modelv1alpha1 "github.com/alibaba/kubedl/apis/model/v1alpha1"
+)
+
+// ModelReconciler reconciles a Model object
+type ModelReconciler struct {
+	client.Client
+	Log    logr.Logger
+	Scheme *runtime.Scheme
+}
+
+var log = logf.Log.WithName("model-controller")
+
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=models,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=model.kubedl.io,resources=models/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=persistentvolumes,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups="",resources=persistentvolumeclaims,verbs=get;list;watch;create;update;patch;delete
+
+// ModelController reconciles and creates the underlying PV and PVC for the model
+func (r *ModelReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+
+	//Get the model
+	model := &modelv1alpha1.Model{}
+	err := r.Get(context.Background(), req.NamespacedName, model)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			log.Info("model doesn't exist", "name", req.String())
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	//TODO move to validation hook
+	if model.Spec.Storage.LocalStorage != nil && model.Spec.Storage.LocalStorage.NodeName == "" {
+		return reconcile.Result{}, fmt.Errorf("local storage should have node name set")
+	}
+
+	pv := &v1.PersistentVolume{}
+	// Does the pv for the model already exist
+	dirty := false
+	err = r.Get(context.Background(), types.NamespacedName{Name: GetModelPVName(model.Name)}, pv)
```

Comment:
is `pv-name-for-model` an implicit and fixed pattern?  what to do when user wants to bind a `Model` with an existed `PV` ?

Reply:
I think user can create a model with the necessary storage info in that case?
Or we can add a volumeName in the model later if needed.